io-sim/src/Control/Monad/IOSim/STM.hs

```diff
@@ -95,7 +95,10 @@ tryPeekTQueueDefault (TQueue queue) = do
       []   -> Nothing
 
 flushTQueueDefault :: MonadSTM m => TQueueDefault m a -> STM m [a]
-flushTQueueDefault (TQueue queue) = uncurry (++) <$> readTVar queue
+flushTQueueDefault (TQueue queue) = do
+  xs <- uncurry (++) <$> readTVar queue
```

Comment:
I think you need to reverse the tail, e.g. 
```suggestion
  xs <- (\front tail -> front ++ reverse tail) <$> readTVar queue
```

Reply:
I know.. I was not trying to fix it yet, but get the other PR merged first.